src/iuse_actor.cpp

```diff
@@ -937,122 +940,89 @@ void firestarter_actor::resolve_firestarter_use( const player *p, const item *,
-            if( skillLevel < 1 ) {
-                // avoid dividing by zero. scaled so that skill level 0 means 60 minutes work
-                skillLevel = 0.536;
-            }
-            // At survival=5 modifier=1, at survival=1 modifier=~6.
-            ///\EFFECT_SURVIVAL speeds up fire starting
-            float moves_modifier = std::pow( 5 / skillLevel, 1.113 );
-            if (moves_modifier < 1) {
-                moves_modifier = 1; // activity time improvement is capped at skillevel 5
-            }
-            const int turns = int( moves_base * moves_modifier );
-            p->add_msg_if_player(m_info, _("At your skill level, it will take around %d minutes to light a fire."), turns / 1000);
-            p->assign_activity(ACT_START_FIRE, turns, -1, p->get_item_position(it), it->tname());
-            p->activity.placement = pos;
-            p->practice( skill_survival, 10 );
-            it->charges -= it->type->charges_to_use() * round(moves_modifier);
-            return 0;
-        }
+    if( t ) {
+        return 0;
     }
-    return 0;
-}
 
-bool extended_firestarter_actor::can_use( const player* p, const item* it, bool t, const tripoint& pos ) const
-{
-    if( !firestarter_actor::can_use( p, it, t, pos ) ) {
-        return false;
+    tripoint pos = spos;
+    float light = light_mod( pos );
+    if( light <= 0.0f ) {
+        p->add_msg_if_player( _("You need direct sunlight to light a fire with this.") );
+        return 0;
     }
 
-    if( need_sunlight ) {
-        return ( g->weather == WEATHER_CLEAR || g->weather == WEATHER_SUNNY ) &&
-                 g->natural_light_level( pos.z ) >= 60 && !g->m.has_flag( TFLAG_INDOORS, pos );
+    if( !prep_firestarter_use( p, it, pos ) ) {
+        return 0;
     }
 
-    return true;
+    double skill_level = p->get_skill_level( skill_survival );
+    ///\EFFECT_SURVIVAL speeds up fire starting
+    float moves_modifier = std::pow( 0.8, std::min( 5.0, skill_level ) );
+    const int moves_base = moves_cost_by_fuel( pos );
+    const int min_moves = std::min<int>( moves_base, sqrt( 1 + moves_base / MOVES( 1 ) ) * MOVES( 1 ) );
+    const int moves = std::max<int>( min_moves, moves_base * moves_modifier ) / light;
+    if( moves > MOVES( MINUTES( 1 ) ) ) {
+        // If more than 1 minute, inform the player
+        static const std::string sun_msg =
+            _("If the current weather holds, it will take around %d minutes to light a fire.");
+        static const std::string normal_msg =
+            _("At your skill level, it will take around %d minutes to light a fire.");
+        p->add_msg_if_player( m_info, ( need_sunlight ? sun_msg : normal_msg ).c_str(),
+            moves / MOVES( MINUTES( 1 ) ) );
+    } else if( moves < MOVES( 2 ) ) {
+        // If less than 2 turns, don't start a long action
+        resolve_firestarter_use( p, it, pos );
+        p->mod_moves( -moves );
+        return it->type->charges_to_use();
```

Comment:
If you're using `ammo_capacity()` and `ammo_remaining()` you may as well use `ammo_required()`